network/limitlistener/rejectingLimitListener.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2013 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// RejectingLimitListener is a modification of LimitListener in
```

Comment:
Would it work to implement this by wrapping `LimitListener` instead of forking it?

```
type rejectingLimitListener struct {
	listener LimitListener
}

func (l *rejectingLimitListener) Accept() (net.Conn, error) {
	select {
	case conn, err := <-l.listener.Accept()
		return conn, err
	default:
		// If the connection is limited, close it immediately instead of waiting.
		err = c.Close()
		if err != nil {
			return nil, fmt.Errorf("Accept() close err: %w", err)
		}
		return nil, errors.New("Accept() limit listener is closed")
	}
}

func (l *rejectingLimitListener) Close() error {
	return l.listener.Close()
}
```

Reply:
I think it wouldn't work. `LimitListener` will not give new connections above the limit, so I will not be able to close them.

Reply:
Why does it need to give a connection?

The `LimitListener` would be stuck inside this select so calling done would unblock the limit listener:
```
func (l *limitListener) acquire() bool {
	select {
	case <-l.done:
		return false
	case l.sem <- struct{}{}:
		return true
	}
}
```

I guess that you're saying we need to `Accept` each connection before closing it? That's already handled by the LimitListener, when acquiring returns false all connections are accepted then closed:
```
func (l *limitListener) Accept() (net.Conn, error) {
	if !l.acquire() {
		// If the semaphore isn't acquired because the listener was closed, expect
		// that this call to accept won't block, but immediately return an error.
		// If it instead returns a spurious connection (due to a bug in the
		// Listener, such as https://golang.org/issue/50216), we immediately close
		// it and try again. Some buggy Listener implementations (like the one in
		// the aforementioned issue) seem to assume that Accept will be called to
		// completion, and may otherwise fail to clean up the client end of pending
		// connections.
		for {
			c, err := l.Listener.Accept()
			if err != nil {
				return nil, err
			}
			c.Close()
		}
	}
```

Reply:
Are you proposing to call `limitListener.Close()`? That would close the underlying listener, we don't want that.

Reply:
@tolikzinovyev thanks, I was missing that point. Sounds like forking the limit listener is the way to go.